base/runtime_internals.jl

```diff
@@ -1182,6 +1182,37 @@ function signature_type(@nospecialize(f), @nospecialize(argtypes))
     return rewrap_unionall(Tuple{ft, u.parameters...}, argtypes)
 end
 
+"""
+    delete_method(m::Method)
+
+Make method `m` uncallable and force recompilation of any methods that use(d) it.
+"""
+function delete_method(m::Method)
+    ccall(:jl_method_table_disable, Cvoid, (Any, Any), get_methodtable(m), m)
+end
+
+"""
+    freeze!(mt::Core.MethodTable)
+
+Disallow adding or modifying methods of `mt`.
+
+See also [`unfreeze!(mt)`](@ref unfreeze!).
+"""
+function freeze!(mt::Core.MethodTable)
```

Comment:
This should perhaps go into Experimental, but anyways, I think a better long-term API is going to be specifying a method `m` which disallows adding any new methods if `m` is not morespecific than them (e.g. asserting that `m` must be the morespecific method whenever it matches). We can approximate that by setting the MethodTable bit, but longer term, the MethodTable object may get removed since it is too gross of an approximation of this (c.f. how non-function-keyword objects and all constructors share the same MethodTable)
```suggestion
function morespecific!(m::Method)
    mt = get_methodtable(m)
    # check that all Methods in this table are morespecific than m
    # so we might avoid disabling a table that might get used for more than just subsets of m
    all(m2 -> m === m2 || morespecific(m2, m), MethodList(mt)) || error("unsupported Method to disable")
```